### PR TITLE
dialects: Add AllReduceOp & YieldOp to GPU.

### PR DESCRIPTION
**Pull request by PapyChacal**

### CODE REVIEW

**Review by webmiche — COMMENTED** [review]
LGTM. General question: Why do you always use the `.mlir` extension? Do you not like xdsl IR? xD

I think using both would be a better testing approach.

File: xdsl/dialects/gpu.py [hunk 1]
```diff
@@ -246,12 +297,49 @@ def get(dim: DimensionAttr) -> ThreadIdOp:
...
+
 #_GPUAttr has to be registered instead of DimensionAttr and AllReduceOperationAttr here.
 # This is a hack to fit MLIR's syntax in xDSL's way of parsing attributes, without making GPU builtin.
 # Hopefully MLIR will parse it in a more xDSL-friendly way soon, so all that can be factored in proper xDSL
 # atrributes.
 GPU = Dialect([
-    BarrierOp, BlockDimOp, BlockIdOp, GlobalIdOp, GridDimOp, LaneIdOp,
-    ModuleOp, ModuleEndOp, NumSubgroupsOp, SetDefaultDeviceOp, SubgroupIdOp,
-    SubgroupSizeOp, ThreadIdOp
+    AllReduceOp,
```
Comment (superlopuh) [inline xdsl/dialects/gpu.py:330]: 🙏

**Review by superlopuh — APPROVED** [review]
LGTM!

**Comment by PapyChacal** [comment]
> LGTM. General question: Why do you always use the `.mlir` extension? Do you not like xdsl IR? xD
> 
> I think using both would be a better testing approach.

AFAIU, xDSL syntax support is to be dropped soon, so I don't write specific tests indeed! Did I misunderstand that?

And I like xDSL syntax better, but prioritize smooth interoperability. Having both would be nice for sure!

**Comment by webmiche** [comment]
> AFAIU, xDSL syntax support is to be dropped soon, so I don't write specific tests indeed! Did I misunderstand that?
> 
> And I like xDSL syntax better, but prioritize smooth interoperability. Having both would be nice for sure!

Yes, but Mathieu has been saying that for almost a year now, so "soon" might be relative 😉 
I am fine with not having tests for xdsl syntax 👍

**Comment by PapyChacal** [comment]
Okay, I would like to avoid doing it for now then, and focus on getting a working GPU dialect for now.

If we realize that we want to support the xDSL syntax more than I understood, I'm happy to add tests for it on a later PR!

**Review by webmiche — APPROVED** [review]

**Pull request merged**